### ISSUE TITLE
Downloading process data hits Lambda Limits

### ISSUE THREAD

**Issue by vdiezel** [ospin-web-dev/hambda#296]
I tried to download a process that ran for days and even with 30 seconds of timeout we ran into a timeout issue. Also it already used 1600 MB of RAM! We need to find some solution that streams or chunks the data...

![Bildschirmfoto von 2020-03-18 08-01-09](https://user-images.githubusercontent.com/37934155/76934069-bc5df200-68ee-11ea-897d-8ed1140316fb.png)

**Comment by O-Susanoo** [comment]
This issue is a very interesting one because it hides two issues i will detail both of them but let's start by describing highlighting both of them

- RAM Issue
- Space issue

Previously when we wanted to create the csv file that contain all process data we went through these steps : 

- fetch all process data from mongodb 
- process data by sorting it first and coping it to a new buffer
- copy again that data to even newer buffer 
- you guess it copying again that data to a newer buffer for processing reasons
- finally call csvWriter to write that file
- read that file again 
- submit that file to S3

as you can clearly see we are copying the same data many times , these data can be big which potentially create a lack of ram even on my own laptop i had issues and i have 8gb , and i think that part of it is due to node being not so good also at handling a lot of data in memory

**Fix RAM issue**
You can check the fix at this branch **_fix-download-processData_** , but remember that the fix is PoC and still need to be refactored

Basically the main idea is to use data only once , and delegate the sorting to mongodb instead of doing it ourselves and processing data by streams & chunks not all of it at once and to do that we do this

- for each stream we bring the list of ids of datapoints but sorted in order
- we start by loading about 50 datapoints into our streams
- and with a while loop we iterate over these data and each time create a new row and write it to a file stream thus the process of reading and writing is done at same time , to do that i had to use this package csv-write-stream which allow me to interact with streams (by stream i mean the dev term not ours) and whenever a ospin datastream is emptied we load it again , if we hit the limit of our datapoints then we just return empty string , the reason why it's like this it's because we need the csv file to have same columns

**Fix Space issue**
One everything was implemented one may think that it will work which does locally at least but not on aws because this time we hit our space limit of the lambda function but at least this time it also work locally , so now we have another issue on hand this one was solved using what's called a PassThrough stream , which is basically a double end stream you write stuff at one end and you get it from the other end , one may ask how ? which is a good question and the answer is s3.upload function provided by aws sdk this function allow to upload a stream into S3 bucket , so the basic idea is that we create a  PassThrough stream we give one end to the s3.upload function and we we write on the other end. this actually solved the issue

**Still to be done** 
After this amazing journey we think nothing could came up but it does this time nothing we can solve smartly it's aws gateway timeout , it seems like we cannot hold the request for a file download over 30 seconds which  is too low to fetch and create a big file that can sometimes become huge , viktor suggested to use SNS to deliver the link to user once the operation is done.

there is also an issue that i should highlight in order to test this code i download a 4.1mb file using both this code and the old code , however i found out that there are a couple of missing rows using this new code not too many but it should be investigated in order to understand why
